File: VB/WpfApplication1/GridColumnValidationHelper.vb
Imports System.Collections.Generic
Imports System.Windows.Controls
Imports System.Windows
Imports DevExpress.Xpf.Grid
Imports System.Globalization

Namespace WpfApplication1

    Public Class ValidationRulesCollection
        Inherits List(Of ValidationRule)

    End Class

    Public Module GridColumnValidationHelper

        Public Function GetValidationRules(ByVal obj As GridColumn) As ValidationRulesCollection
            Return CType(obj.GetValue(ValidationRulesProperty), ValidationRulesCollection)
        End Function

        Public Sub SetValidationRules(ByVal obj As GridColumn, ByVal value As ValidationRulesCollection)
            obj.SetValue(ValidationRulesProperty, value)
        End Sub

        Public ReadOnly ValidationRulesProperty As DependencyProperty = DependencyProperty.RegisterAttached("ValidationRules", GetType(ValidationRulesCollection), GetType(GridColumnValidationHelper), New PropertyMetadata(Nothing, AddressOf OnValidationRulesChanged))

        Private Sub OnValidationRulesChanged(ByVal d As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
            Dim column As GridColumn = TryCast(d, GridColumn)
            If column Is Nothing Then Return
            RemoveHandler column.Validate, New GridCellValidationEventHandler(AddressOf OnColumnValidate)
            If e.NewValue IsNot Nothing Then AddHandler column.Validate, New GridCellValidationEventHandler(AddressOf OnColumnValidate)
        End Sub

        Private Sub OnColumnValidate(ByVal sender As Object, ByVal e As GridCellValidationEventArgs)
            Dim column As GridColumn = CType(sender, GridColumn)
            For Each rule As ValidationRule In GetValidationRules(column)
                Dim result As ValidationResult = rule.Validate(e.Value, CultureInfo.CurrentCulture)
                If Not result.IsValid Then
                    e.IsValid = False
                    e.ErrorContent = result.ErrorContent
                    Exit For
                End If
            Next
        End Sub
    End Module
End Namespace

File: CS/WpfApplication1/GridColumnValidationHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows.Controls;
using System.Windows;
using DevExpress.Xpf.Grid;
using System.Globalization;

namespace WpfApplication1 {
    public class ValidationRulesCollection : List<ValidationRule> { }
    public static class GridColumnValidationHelper {


        public static ValidationRulesCollection GetValidationRules(GridColumn obj) {
            return (ValidationRulesCollection)obj.GetValue(ValidationRulesProperty);
        }

        public static void SetValidationRules(GridColumn obj, ValidationRulesCollection value) {
            obj.SetValue(ValidationRulesProperty, value);
        }

        public static readonly DependencyProperty ValidationRulesProperty =
            DependencyProperty.RegisterAttached("ValidationRules", typeof(ValidationRulesCollection), typeof(GridColumnValidationHelper), new PropertyMetadata(null, OnValidationRulesChanged));

        static void OnValidationRulesChanged(DependencyObject d, DependencyPropertyChangedEventArgs e) {
            GridColumn column = d as GridColumn;
            if(column == null)
                return;
            column.Validate -= new GridCellValidationEventHandler(OnColumnValidate);
            if(e.NewValue != null)
                column.Validate += new GridCellValidationEventHandler(OnColumnValidate);
        }

        static void OnColumnValidate(object sender, GridCellValidationEventArgs e) {
            GridColumn column = (GridColumn)sender;
            foreach(ValidationRule rule in GetValidationRules(column)) {
                ValidationResult result = rule.Validate(e.Value, CultureInfo.CurrentCulture);
                if(!result.IsValid) {
                    e.IsValid = false;
                    e.ErrorContent = result.ErrorContent;
                    break;
                }
            }
             
        }
     
    }
}
